src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProducerExtension.cs

```diff
@@ -184,7 +191,7 @@ internal Task DeliverItem(StreamId streamId, object item, bool fireAndForgetDeli
             {
                 var tasks = fireAndForgetDelivery ? null : new List<Task>();
                 var immutableItem = new Immutable<object>(item);
-                foreach (KeyValuePair<GuidId, Tuple<IStreamConsumerExtension, IStreamFilterPredicateWrapper>> subscriptionKvp in consumers)
+                foreach (KeyValuePair<GuidId, Tuple<IStreamConsumerExtension, IStreamFilterPredicateWrapper>> subscriptionKvp in consumers.ToList())
```

Comment:
I don't like that ToList() copy. Its on a critical pass of every delivery. I understand the need, but I think perf argument here justifies a bit more complicated code. 
I think the easiest is to use ContinuewWith instead.
On the grain its single threaded, so no problem.
On the client (this code running on the client, yes, can happen client delivering to client) the dictionary is concurrent, so will work too.